File: pms/lib/controllers/TeamController.dart
import 'dart:convert';
import 'package:get/get.dart';
import 'package:http/http.dart' as http;
import '../configurations.dart';

class TeamController extends GetxController
{
  // this function will make a post req to create a team
  static Future<void> createTeam(String teamName, Set<String>teamMembers) async {
    var url = Uri.parse('${Configuration.apiBaseUrl}team/create');
    var memberIds = json.encode(teamMembers.toList());
    try {
      final response = await http.post(
        url,
        body: {
          'title': teamName,
          'member_id': memberIds,
        },
      );

      if (response.statusCode == 200) {
        print(json.decode(response.body));

        var responseData = json.decode(response.body);
        print(responseData['code']);
        Get.back();
      }
    } catch (error) {
      print(error);
    }
  }

  // this function will return List of all the teams
  static Future<List<Map<String, dynamic>>> getTeams() async {
    var url = Uri.parse('${Configuration.apiBaseUrl}team/view');
    List<Map<String, dynamic>> teamsList = []; // List of maps
    try {
      final response = await http.get(url);
      if (response.statusCode == 201) {
        List<dynamic> teamResponse = json.decode(response.body);
        teamResponse.forEach((result) {
          teamsList.add({
            "team_id":result['_id'],
            "title": result['title'],
            // "member_ids": result['member_id']
            //     .map((member) => member['_id'])
            //     .toList(),
          });
        });
      }
    } catch (error) {
      print(error);
    }
    return teamsList;
  }

  // this function will return details of a specific team
  static Future<Map> getTeambyId(var id) async {
    Map team = {
      "title": '',
    };
    var url = Uri.parse('${Configuration.apiBaseUrl}team/view');
    try {
      final response = await http.get(url);
      if (response.statusCode == 201) {
        Map<String,dynamic> teamResponse = json.decode(response.body);
        Map team = {
          "title": teamResponse['title'].toString(),
        };
      }
    } catch (error) {
      print(error);
    }
    return team;
  }

  // this function will return teams in which the member is present
  // member id is given to this function
  static Future<List<Map<String, dynamic>>> getTeamByMemberId(var id) async {
    var url = Uri.parse('${Configuration.apiBaseUrl}team/getbymember/'+id.toString());
    List<Map<String, dynamic>> teamsList = []; // List of maps
    try {
      final response = await http.get(url);
      if (response.statusCode == 200) {
        List<dynamic> teamResponse = json.decode(response.body);
        print(teamResponse);
        teamResponse.forEach((result) {
          teamsList.add({
            "team_id":result['_id'],
            "title": result['title'],
            // "member_ids": result['member_id']
            //     .map((member) => member['_id'])
            //     .toList(),
          });
        });
      }
    } catch (error) {
      print(error);
    }
    return teamsList;
  }




}

File: backend/controllers/TeamController.js
const Team = require('../models/Team');
const LogController = require('../controllers/LogController');

// this function will create a team after given the title and members id
async function createTeam(req,res){
    console.log(req.body);
    try{
        req.body.member_id = JSON.parse(req.body.member_id);
        const team = await Team.create(req.body);
        res.status(201).json(team);
    }
    catch(err){
        LogController.createLog("TeamController.js","createTeam",err.message);
        res.status(500).json({error: err.message});
    }
}
// this function will return all teams
async function getAllTeams(req,res){
    try{
        const teams = await Team.find();
        console.log(teams);
        res.status(201).json(teams);
    }
    catch(err){
        LogController.createLog("TeamController.js","getAllTeams",err.message);
        res.status(500).json({error:err.message});
    }
}

//this function will return team details of a specific member
async function getTeamByMemberId(req, res) {
    try {
        const { id } = req.params;
        const teams = await Team.find({ member_id: id });
        res.status(200).json(teams);
    } catch (err) {
        LogController.createLog("TeamController.js","getTeamByMemberId",err.message);
        res.status(500).json({ error: err.message });
    }
}


async function updateTeams(req, res) {
    try {
        const { id } = req.params;
        const updates = req.body;
        const updatedTeam = await Project.findOneAndUpdate({ id: id }, updates, { new: true });
        
        res.status(201).json(updatedTeam);
    } catch (err) {
        LogController.createLog("TeamController.js","updateTeams",err.message);
        res.status(500).json({ error: err.message });
    }
}

// this function will delete the team after given the team id
async function deleteTeam(req,res){
    try{
        const { id } = req.params;
        await Team.findOneAndRemove({ id: id });
        res.sendStatus(201);
    }
    catch(err){
        LogController.createLog("TeamController.js","deleteTeam",err.message);
        res.status(500).json({error: err.message});
    }
}

module.exports={
    createTeam,
    getAllTeams,
    updateTeams,
    deleteTeam,
    getTeamByMemberId
};